File: src/solar/util/ability/AbilityMappingUtil.ts
import LangUtil from "../lang/LangUtil";

const CJ: NoSelf = require('jass.common')
/**
 * 技能映射
 *
 * 通常在使用QWER模板技能时 释放自定义的技能
 *
 */
export default class AbilityMappingUtil {

    private static abilityMappingData: {
        [unitHandleId: number]: {
            [qwerId: number]: number
        }
    } = {}


    /**
     * 将类似qwer这种模板id 在释放时 映射到真正需要释放的技能id
     * @param unit 需要映射技能的 单位
     * @param templateId qwer模板的技能物编id
     * @param targetId 目标技能id也可以是虚拟的技能(即无真实技能物编是这个id)
     */
    static mapping(unit: unit, templateId: string, targetId: string) {
        //启动映射系统
        AbilityMappingUtil._sl_init_mapping_system();
        //
        let unitHandleId = GetHandleId(unit);
        let mappingDatum = AbilityMappingUtil.abilityMappingData[unitHandleId];
        if (!mappingDatum) {
            AbilityMappingUtil.abilityMappingData[unitHandleId] = {};
        }
        mappingDatum[FourCC(templateId)] = FourCC(targetId);

    }


    /**
     *
     * @param unit 移除所有技能映射的单位
     */
    static removeMappings(unit: unit): void {
        let unitHandleId = GetHandleId(unit);
        let mappingDatum = AbilityMappingUtil.abilityMappingData[unitHandleId];
        if (mappingDatum) {
            AbilityMappingUtil.abilityMappingData[unitHandleId] = null;
        }
    }

    /**
     *
     * @param unit 移除映射的单位
     * @param templateId 指定映射的技能
     */
    static removeMapping(unit: unit, templateId: string): void {
        let unitHandleId = GetHandleId(unit);
        let mappingDatum = AbilityMappingUtil.abilityMappingData[unitHandleId];
        if (mappingDatum) {
            mappingDatum[FourCC(templateId)] = null;
        }
    }

    /**
     *
     * 获取真实的释放id 获取到模板id 即没有经过本系统处理的id
     */
    static getRealSpellAbilityId() {
        return CJ.GetSpellAbilityId();
    }


    private static _sl_isInitialized = false

    private static _sl_init_mapping_system() {
        if (AbilityMappingUtil._sl_isInitialized) {
            return;
        }
        //重写 GetSpellAbilityId
        let oldGetSpellAbilityId = GetSpellAbilityId;
        _G.GetSpellAbilityId = function (this: void) {
            let cjRealId = oldGetSpellAbilityId();
            //判断是否有映射的id
            let unitHandleId = GetHandleId(GetTriggerUnit());
            let targetId = AbilityMappingUtil.abilityMappingData[unitHandleId]?.[cjRealId];
            if (targetId) {
                //返回映射后的id
                return targetId;
            }
            //返回真实的id
            return cjRealId;
        }
        //重写 GetUnitAbilityLevel 以应对被动技能
        let oldGetUnitAbilityLevel = GetUnitAbilityLevel;
        _G.GetUnitAbilityLevel = function (this: void, whichUnit: unit, abilcode: number | string) {

            //判断是否有映射的id
            let unitHandleId = GetHandleId(whichUnit);
            if (LangUtil.isString(abilcode)) {
                abilcode = FourCC(abilcode as string)
            }
            //遍历是否有此id对应的模板id
            let mappingDatum = AbilityMappingUtil.abilityMappingData[unitHandleId];
            for (let templateIdNum in mappingDatum) {
                if (mappingDatum[templateIdNum] == abilcode) {
                    //返回映射的模板id等级
                    return oldGetUnitAbilityLevel(whichUnit, templateIdNum);
                }
            }
            //返回真实的等级
            let cjRealLv = oldGetUnitAbilityLevel(whichUnit, abilcode);
            return cjRealLv;
        }

        //
        AbilityMappingUtil._sl_isInitialized = true;
    }

}

File: scripts/solar/util/ability/AbilityMappingUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 1,["7"] = 1,["8"] = 3,["9"] = 10,["10"] = 10,["11"] = 10,["13"] = 10,["14"] = 25,["15"] = 27,["16"] = 29,["17"] = 30,["18"] = 31,["19"] = 32,["21"] = 34,["22"] = 25,["23"] = 43,["24"] = 44,["25"] = 45,["26"] = 46,["27"] = 47,["29"] = 43,["30"] = 56,["31"] = 57,["32"] = 58,["33"] = 59,["34"] = 60,["36"] = 56,["37"] = 68,["38"] = 69,["39"] = 68,["40"] = 75,["41"] = 76,["44"] = 80,["45"] = 81,["46"] = 82,["47"] = 84,["48"] = 85,["49"] = 85,["50"] = 85,["52"] = 85,["53"] = 86,["54"] = 88,["56"] = 91,["57"] = 81,["58"] = 94,["59"] = 95,["60"] = 98,["61"] = 99,["62"] = 100,["64"] = 103,["65"] = 104,["66"] = 105,["67"] = 107,["70"] = 111,["71"] = 112,["72"] = 95,["73"] = 116,["74"] = 75,["75"] = 16,["76"] = 73,["77"] = 10});
local ____exports = {}
local ____LangUtil = require("solar.util.lang.LangUtil")
local LangUtil = ____LangUtil.default
local CJ = require("jass.common")
____exports.default = __TS__Class()
local AbilityMappingUtil = ____exports.default
AbilityMappingUtil.name = "AbilityMappingUtil"
function AbilityMappingUtil.prototype.____constructor(self)
end
function AbilityMappingUtil.mapping(self, unit, templateId, targetId)
    ____exports.default:_sl_init_mapping_system()
    local unitHandleId = GetHandleId(unit)
    local mappingDatum = ____exports.default.abilityMappingData[unitHandleId]
    if not mappingDatum then
        ____exports.default.abilityMappingData[unitHandleId] = {}
    end
    mappingDatum[FourCC(templateId)] = FourCC(targetId)
end
function AbilityMappingUtil.removeMappings(self, unit)
    local unitHandleId = GetHandleId(unit)
    local mappingDatum = ____exports.default.abilityMappingData[unitHandleId]
    if mappingDatum then
        ____exports.default.abilityMappingData[unitHandleId] = nil
    end
end
function AbilityMappingUtil.removeMapping(self, unit, templateId)
    local unitHandleId = GetHandleId(unit)
    local mappingDatum = ____exports.default.abilityMappingData[unitHandleId]
    if mappingDatum then
        mappingDatum[FourCC(templateId)] = nil
    end
end
function AbilityMappingUtil.getRealSpellAbilityId(self)
    return CJ.GetSpellAbilityId()
end
function AbilityMappingUtil._sl_init_mapping_system(self)
    if ____exports.default._sl_isInitialized then
        return
    end
    local oldGetSpellAbilityId = GetSpellAbilityId
    _G.GetSpellAbilityId = function()
        local cjRealId = oldGetSpellAbilityId()
        local unitHandleId = GetHandleId(GetTriggerUnit())
        local ____AbilityMappingUtil_abilityMappingData_unitHandleId_cjRealId_0 = ____exports.default.abilityMappingData[unitHandleId]
        if ____AbilityMappingUtil_abilityMappingData_unitHandleId_cjRealId_0 ~= nil then
            ____AbilityMappingUtil_abilityMappingData_unitHandleId_cjRealId_0 = ____AbilityMappingUtil_abilityMappingData_unitHandleId_cjRealId_0[cjRealId]
        end
        local targetId = ____AbilityMappingUtil_abilityMappingData_unitHandleId_cjRealId_0
        if targetId then
            return targetId
        end
        return cjRealId
    end
    local oldGetUnitAbilityLevel = GetUnitAbilityLevel
    _G.GetUnitAbilityLevel = function(whichUnit, abilcode)
        local unitHandleId = GetHandleId(whichUnit)
        if LangUtil:isString(abilcode) then
            abilcode = FourCC(abilcode)
        end
        local mappingDatum = ____exports.default.abilityMappingData[unitHandleId]
        for templateIdNum in pairs(mappingDatum) do
            if mappingDatum[templateIdNum] == abilcode then
                return oldGetUnitAbilityLevel(whichUnit, templateIdNum)
            end
        end
        local cjRealLv = oldGetUnitAbilityLevel(whichUnit, abilcode)
        return cjRealLv
    end
    ____exports.default._sl_isInitialized = true
end
AbilityMappingUtil.abilityMappingData = {}
AbilityMappingUtil._sl_isInitialized = false
____exports.default = AbilityMappingUtil
return ____exports
